src/server/multiplayer/MultiplayerServerState.js

```diff
@@ -0,0 +1,164 @@
+/* eslint-disable no-console */
+import { compact, find, pull, reject, without } from 'lodash';
+
+import { id as generateID } from '../../common/util/common';
+
+// Returns a copy of a game with the given client
+// removed from both the players and spectators lists (if present),
+// or null if this results in the game no longer having >=2 active players.
+function withoutClient(game, clientID) {
+  const updatedGame = Object.assign({}, game, {
+    players: without(game.players, clientID),
+    spectators: without(game.spectators, clientID)
+  });
+  return updatedGame.players.length >= 2 ? updatedGame : null;
+}
+
+export default class MultiplayerServerState {
+  state = {
+    connections: {},  // map of { clientID: websocket }
+    games: [],  // array of { id, name, players, playerColors, spectators, actions, decks, usernames, startingSeed }
+    waitingPlayers: [],  // array of { id, name, deck, players }
+    playersOnline: [],  // array of clientIDs
+    usernames: {}  // map of { clientID: username }
+  };
+
+  /* Getters */
+
+  // Returns a serializable subset of the state for broadcast as an INFO message.
+  serialize = () => {
+    const { games, waitingPlayers, playersOnline, usernames } = this.state;
+    return { games, waitingPlayers, playersOnline, usernames };
+  }
+
+  // Returns the socket corresponding to a given player.
+  getClientSocket = (clientID) => (
+    this.state.connections[clientID]
+  )
+
+  // Returns all websocket connections corresponding to the given clientIDs,
+  // or ALL connections if no clientIDs are specified.
+  getClientSockets = (clientIDs = null) => (
+    clientIDs ? clientIDs.map(this.getClientSocket) : Object.values(this.state.connections)
+  )
+
+  // Returns the username to use for the given player.
+  // If fallbackToClientID is true, falls back to the client ID if there is no
+  // username set, otherwise returns null.
+  getClientUsername = (clientID, fallbackToClientID = true) => (
+    this.state.usernames[clientID] || (fallbackToClientID && clientID)
+  )
+
+  // Returns the game that the given player is in, if any.
+  lookupGameByClient = (clientID) => (
+    this.state.games.find(game => game.players.includes(clientID))
+  )
+
+  // Returns all *other* players in the game that the given player is in, if any.
+  getAllOpponents = (clientID) => {
+    const game = this.state.games.find(g => [...g.players, ...g.spectators].includes(clientID));
+    if (game) {
+      return [...game.players, ...game.spectators].filter(id => id !== clientID);
```

Comment:
Maybe `[...game.players, ...game.spectators]` can be extracted to a util function like `getPeopleInGame`?